libs/sdk-core/src/breez_services.rs

```diff
@@ -574,6 +574,13 @@ impl BreezServices {
         Ok(self.persister.get_payment_by_hash(&hash)?)
     }
 
+    /// Set the external metadata of a payment as a valid JSON string
+    pub async fn set_payment_metadata(&self, hash: String, metadata: String) -> SdkResult<()> {
```

Comment:
I still don't see json validation and even if we do so what happen if the user uses a nested json? Do we allow this? Won't it break the search?
How about passing here an array of metadata where every metadata has `key` and `value` fields where both are strings? We can validate for duplicate keys and it makes the search a lot simpler since we only use strings as values.
@dangeross what do you think?

Reply:
I added the JSON validation inside the `update_payment_external_metadata` method, see [here](https://github.com/hydra-yse/breez-sdk/blob/febab11e462fe20d8f334eaf5e178711150bcc68/libs/sdk-core/src/persist/transactions.rs#L110-L111). Regarding passing a vector of `PaymentMetadata`, I agree, since we can then also reduce everything to one SQL query.

Reply:
As @hydra-yse says, you can query within the nested JSON using `json_extract` with a key like `$.parent.child`, even with arrays `$.parent.arr[*].child`. I don't have a problem with the metadata being JSON, I think the question is what is the best approach to use to sync the metadata. A key/value store in a separate table is probably a better approach to keep the payment syncing clean. With a separate table when syncing we can control the approach to updating we want to take based on the metadata payment_hash/key, be it write once or write many (based on timestamp).

Reply:
I agree a separate table is the correct approach as the sync strategy is different.
My initial thinking was one row per payment with a json field.
When you referring to key/value store do you mean one row per metadata key so there is one to many relation between payment to metadata? This will complicate the interface as we will need to support also delete.

Reply:
Ah yeah, I was thinking one to many with one row per key/value but didn't consider the delete case. Then one row per payment_hash/json seems more appropriate.

Reply:
With JSON, the question of validation remains. AFAIK duplicate keys are valid in JSON, but would be invalid in our scenario. So we'd have to explicitly check and deny that.